File: 206-Reverse-Linked-List/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if(head==NULL || head->next==NULL) return head;
        ListNode *preNode, *curNode, *nextNode;
        preNode = head;
        curNode = head->next;
        head->next = NULL;
        while(curNode) {
            nextNode = curNode->next;
            curNode->next = preNode;
            preNode = curNode;
            curNode = nextNode;
        }
        return preNode;
    }
};

File: 206-Reverse-Linked-List/solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    // T: O(n)
    // S: O(1)
    public ListNode reverseList(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }

        ListNode prev = head;
        head = head.next;
        ListNode next = head.next;

        prev.next = null;
        head.next = prev;

        while (next != null) {
            prev = head;
            head = next;
            next = next.next;
            head.next = prev;
        }

        return head;
    }
}